python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1897,3 +1896,108 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
     """
     sorted_key_strs = sorted(str(key) for key in entity_keys)
     return non_secure_md5_hash_str(json.dumps(sorted_key_strs).encode("utf-8"))[:8]
+
+
+def update_asset_in_mappings_on_node(
+    assets_def: AssetsDefinition, updated_deps: Mapping[AssetKey, AbstractSet[AssetDep]]
+) -> "AssetsDefinition":
+    """This function threads additional AssetDep objects into the node definition of the
+    AssetsDefinition. It assumes that the deps have already been added to the AssetSpecs of the AssetsDefinition, and purely handles mapping those deps to inputs on the op.
+    """
+    from dagster._builtins import Nothing
+    from dagster._core.definitions.input import In
+
+    # If the assets_def is not executable (and therefore is not backed by a node definition), return the original assets_def
+    if not assets_def.is_executable:
+        return assets_def
+    original_key_to_input_mapping = reverse_dict(assets_def.node_keys_by_input_name)
+    # Get the new deps that are not already in the original mapping
+    new_deps = {
```

Comment:
This code is a bit hard to follow, which is partly unavoidable as we have all these annoying dictionaries, but the way I'm thinking about it, it might be easier to handle this purely as sets to start with, i.e.:

```python
# mapped_specs is the list of specs that were mapped in the above fn
new_deps = set().union(*(spec.deps for spec in mapped_specs))
previous_deps = set().union(*(spec.deps for spec in original_assets_def.specs))

added_deps = new_deps-previous_deps
deleted_deps = previous_deps-new_deps

if not added_deps and not deleted_deps: ...
```

Possible that I'm missing some subtle or not-so-subtle stuff that makes that pattern not work, so if that's the case feel free to ignore

Reply:
quality suggestion. I think the core complication comes from me trying to save a linear pass over the replaced specs. Which is .... dum